### ISSUE TITLE
Unclear when working

### ISSUE THREAD

**Issue by ChildishGiant** [UB-Mannheim/zotero-ocr#54]
The addon gives the user no feedback that it's actually doing anything, potentially leading to action spamming and then making multiple copies of everything

**Comment by kaiwen-wang** [comment]
select 'show file' to view folder progress

**Comment by v4u6h4n** [comment]
Thanks for making this plugin, its so helpful to have this feature integrated into zotero :-) However I would like to request that a progress indicator be implemented, as checking the files folder is a bit of a workflow disruption. Should I make a separate issue for the request, or is just mentioning it here enough?

**Comment by aborel** [comment]
Stating your interest here is sufficient, thanks!

**Comment by aborel** [comment]
I have a partial solution, I'm working on it.

**Comment by v4u6h4n** [comment]
Oh nice! :-)

**Comment by aborel** [comment]
I'm considering a progress window with some status information, similar to the one for the browser connector ( https://i.sstatic.net/mHh7x.jpg for a quick example randomly found in the forum).
Ideally I'd like a count of pages processed but I'm not sure it's feasible, the execution of tesseract is kind of a black box for monitoring: there's apparently no way to capture tesseract messages in real time. What I can do instead is display a timer to inform the user that the script is still running. Does that sound acceptable?

**Comment by stweil** [comment]
Tesseract is called with a list of image files (one per page of the original PDF file). It prints a message for each processed page. Example for a list with 10 JPEG images:

```
Page 0 : image1.jpg
Page 1 : image2.jpg
[...]
Page 9 : image10.jpg
```

Maybe this message output can be captured and used for a progress display.

**Comment by aborel** [comment]
There's the problem :-/ So far, I haven't found a way to capture the message output. Zotero.Utilities.Internal.exec() doesn't do it. A simple redirection doesn't really work due to buffering: lines will not appear in real time, we must wait for the system buffer to be filled before we get a whole batch of lines (about 100 lines for a 1-kb buffer, that's a long time without any user information).

I'm not very optimistic after reading https://forums.zotero.org/discussion/79157/capture-stdout-from-shell-script but things might have changed since 2019. I'm still searching for a possible solution.

One hypothetical strategy would be to run tesseract on 1 image at a time instead of a list, but then we would need to figure out a way to reassemble the actual OCR output.

**Comment by aborel** [comment]
One possible strategy would be to run tesseract one image at a time (which we can track easily enough)  and assemble the multiple PDF outputs using Poppler's `pdfunite` tool. Managing the other text outputs feels straightforward.
It would probably facilitate a solution to https://github.com/UB-Mannheim/zotero-ocr/issues/72 , but it takes us one more step away from https://github.com/UB-Mannheim/zotero-ocr/issues/80 .

@stweil what do you think?

**Comment by aborel** [comment]
@stweil Any thoughts about the pdfunite suggestion?

**Comment by stweil** [comment]
Using single pages and pdfunite would work, but make things more complicated and possibly add some overhead which would result in larger PDF files.

Finding a solution which captures the output from `tesseract` (which should print its messages without buffering) would be more elegant, but up to now I had no time to try this.

**Comment by aborel** [comment]
I agree, but I haven't found any way to do that. That's why I'm suggesting this route.

**Comment by aborel** [comment]
@auscyber thanks for https://forums.zotero.org/discussion/125991/pull-continual-output-from-executed-program , this looks more promising than anything I have read so far.

**Comment by auscyber** [comment]
I’m gonna make a proof of concept and see if we can pull continual output but i’m not an expert zotero ui dev so i’m unsure how to do progress windows so ill just make it print to console

Thanks
Ivy
________________________________

**Comment by aborel** [comment]
Don't worry too much about the UI, I'm confident that a successful proof of concept could be integrated into our relevant branch https://github.com/UB-Mannheim/zotero-ocr/tree/progressbar

**Comment by v4u6h4n** [comment]
Not sure if it's possible, but it might be cool to have an imbedded icon showing the process status similar to the native button/status animated icon that shows up when metadata is being retreived for a new item, in the top right hand corner next to the sync button/status icon.

**Comment by aborel** [comment]
That's not on the radar right now, but perhaps we could consider it later. Thanks for the suggestion!

**Comment by aborel** [comment]
769b8ea580081b21bc1a93a92d96fdf2babf870c is almost ready for release. Thank you so much @auscyber !
